src/Particle/ParticleSpatialLayout.hpp

```diff
@@ -162,34 +184,139 @@ namespace ippl {
+
+        return totalSize;
+    }
+
     template <typename T, unsigned Dim, class Mesh, typename... Properties>
     template <typename ParticleContainer>
     detail::size_type ParticleSpatialLayout<T, Dim, Mesh, Properties...>::locateParticles(
         const ParticleContainer& pc, locate_type& ranks, bool_type& invalid) const {
-        auto& positions                            = pc.R.getView();
-        typename RegionLayout_t::view_type Regions = rlayout_m.getdLocalRegions();
+        auto& positions          = pc.R.getView();
+        region_view_type Regions = rlayout_m.getdLocalRegions();
 
         using mdrange_type = Kokkos::MDRangePolicy<Kokkos::Rank<2>, position_execution_space>;
 
-        int myRank = Comm->rank();
+        size_type myRank = Comm->rank();
 
         const auto is = std::make_index_sequence<Dim>{};
 
-        size_type invalidCount = 0;
-        Kokkos::parallel_reduce(
+        const neighbor_list& neighbors = flayout_m.getNeighbors();
+
+        // container of particles that travelled more than one cell
+        locate_type notFoundIds("Not found", size_type(0.1 * pc.getLocalNum()));
+        // Now: dimension hard-coded, for future implementations maybe make it as a run parameter.
+        bool_type found("Found", pc.getLocalNum());
+        size_type nLeft              = 0;
+        size_type invalidCount       = 0;
+        const size_type neighborSize = getNeighborSize(neighbors);
+        locate_type neighbors_view("Nearest neighbors IDs", neighborSize);
+
+        increment_type red_val;
+        red_val.init();
+
+        auto neighbors_mirror =
+            Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), neighbors_view);
+
+        size_t k = 0;
+
+        for (const auto& componentNeighbors : neighbors) {
+            for (size_t j = 0; j < componentNeighbors.size(); ++j) {
+                neighbors_mirror(k) = componentNeighbors[j];
+                k++;
+            }
+        }
+
+        Kokkos::deep_copy(neighbors_view, neighbors_mirror);
+
+        /*Begin Kokkos loop:
+         *Step 1: search in current rank
+         *Step 2: search in neighbors
+         *Step 3: save information on whether the particle was located
+         *Step 4: run additional loop on non-located particles*/
+        
+        Kokkos::parallel_scan(
             "ParticleSpatialLayout::locateParticles()",
-            mdrange_type({0, 0}, {ranks.extent(0), Regions.extent(0)}),
-            KOKKOS_LAMBDA(const size_t i, const size_type j, size_type& count) {
-                bool xyz_bool = positionInRegion(is, positions(i), Regions(j));
+            Kokkos::RangePolicy<size_t>(0, ranks.extent(0)),
+            KOKKOS_LAMBDA(const size_type i, increment_type& val, const bool final) {
+                // Step 1
+                bool xyz_bool = false;
+                bool increment[2];
+
+                xyz_bool = positionInRegion(is, positions(i), Regions(myRank));
+
                 if (xyz_bool) {
```

Comment:
I think all these `if`s can be avoided by doing something similar to what we do in other places in IPPL
`ranks(i) = xyz_bool * rank;
invalid(i) = !xyz_bool;
found(i) = xyz_bool;
`
This will put `0` as the rank for the particles which are not found in the current rank. But it is ok since that will be replaced by the correct ranks in the steps 2 and 4. Similar to above the `if`s in steps 2 and 4 can also be avoided.

Reply:
The only doubt i have on removing the `if` is how can i handle the break from the loop

Reply:
But the other ones can be removed, you're right

Reply:
Why do you need to break from that loop? Even if you don't break you are going to test for the other field neighbors (which is not needed) but since this is small (O(10)s only) and GPUs especially don't care about it in my opinion we don't need that break.

Reply:
My question is:

Say `xyz_found` is true for the first neighbor. The function assigns `invalid(i) = xyz_found`, then at the next cycle you get `xyz_found = false.` If you reassign `invalid(i) = xyz_found`, the previous true valued will  be overwritten...

Reply:
> @aaadelmann I usually check what comments are resolved/unresolved and approve based on that. So in future it would be of great help if you could check carefully before marking discussions as resolved. Because I accidentally skipped this one which in my opinion is still unresolved.

Sorry I think this is because of my lack of understanding of the github workflow. However I see that this is not approved yet so I guess I do not understand the issue ... because of my lack of understanding of the github workflow. Lets maybe discuss this on a Zoom so I better get the ideas.

Reply:
@aaadelmann No problem I am also not 100 percent sure with the github workflow and trying to figure out things which were in the gitlab workflow. Yes in this case it is fine as I somehow noticed Veronica's message before and searched through the resolved discussions to find this one. But I usually don't do that and just look at the PR again and see what things are still unresolved (assuming the things which are resolved are ok). Anyway as you mentioned a Zoom may be better also with @matt-frey as he is also more experienced with github workflow.

Reply:
@vmontanaro Yeah didn't think about it before. How about `invalid(i) += xyz_found` would that create a race condition? Since the Kokkos loop is over the particles multiple threads should not write but I am not entirely sure..

Reply:
I could try with that one, it should not give problems.

Reply:
> @vmontanaro Yeah didn't think about it before. How about `invalid(i) += xyz_found` would that create a race condition? Since the Kokkos loop is over the particles multiple threads should not write but I am not entirely sure..

This gives a `Kokkos::Impl::SharedAllocationRecord 'Found' failed decrement count = 0`.

Reply:
> > @vmontanaro Yeah didn't think about it before. How about `invalid(i) += xyz_found` would that create a race condition? Since the Kokkos loop is over the particles multiple threads should not write but I am not entirely sure..
> 
> @srikrrish This gives a `Kokkos::Impl::SharedAllocationRecord 'Found' failed decrement count = 0`.

Fixed in last commit.

Reply:
Yeah this seems to be good too..